indexer/src/main/java/org/uniprot/store/indexer/common/config/UniProtSolrClient.java

```diff
@@ -0,0 +1,160 @@
+package org.uniprot.store.indexer.common.config;
+
+import static java.util.Arrays.asList;
+
+import java.io.IOException;
+import java.util.Collection;
+import java.util.List;
+import java.util.Optional;
+
+import lombok.extern.slf4j.Slf4j;
+
+import org.apache.http.client.HttpClient;
+import org.apache.solr.client.solrj.SolrClient;
+import org.apache.solr.client.solrj.SolrQuery;
+import org.apache.solr.client.solrj.SolrServerException;
+import org.apache.solr.client.solrj.impl.CloudSolrClient;
+import org.apache.solr.client.solrj.impl.HttpClientUtil;
+import org.apache.solr.client.solrj.impl.HttpSolrClient;
+import org.apache.solr.client.solrj.response.QueryResponse;
+import org.apache.solr.client.solrj.response.UpdateResponse;
+import org.apache.solr.common.params.ModifiableSolrParams;
+import org.uniprot.store.search.SolrCollection;
+import org.uniprot.store.search.document.Document;
+
+/**
+ * A wrapper of {@link SolrClient} which creates a {@link SolrClient} instance for each thread that
+ * it is used on. The purpose of this is to ensure that multi-threaded applications have multiple
+ * instances that can access Solr, e.g., improving writing throughput.
+ *
+ * <p>Created 10/07/19
+ *
+ * @author Edd
+ */
+@Slf4j
+public class UniProtSolrClient {
+    private final ThreadLocal<SolrClient> threadLocalSolrClients;
+    private final RepositoryConfigProperties config;
+
+    UniProtSolrClient(RepositoryConfigProperties config) {
+        this.config = config;
+        this.threadLocalSolrClients = ThreadLocal.withInitial(this::uniProtSolrClient);
+    }
+
+    public UniProtSolrClient(SolrClient solrOperations) {
+        this.config = null;
+        this.threadLocalSolrClients = ThreadLocal.withInitial(() -> solrOperations);
+    }
+
+    public <T extends Document> List<T> query(
```

Comment:
really good question. I don't think we can infer it from the return type, because this value comes from Solr's `getBeans(Class)` method, but perhaps we could have a map of collection <-> document, because we do know the collection in the call.

I'll create a jira for this as it's a nice idea. thank you :)